File: scripts/flutter/template/flutter_di_template.py
#!/usr/bin/env python3
"""Flutter DI 등록 템플릿 생성 스크립트

injection_container.dart에 추가할 BLoC 등록 코드를 생성합니다.
"""

import sys


def to_snake_case(name: str) -> str:
    """PascalCase를 snake_case로 변환"""
    result = []
    for i, char in enumerate(name):
        if char.isupper() and i > 0:
            result.append("_")
        result.append(char.lower())
    return "".join(result)


def generate_di_registration(screen_name: str, feature_dir: str) -> str:
    """DI 등록 코드 생성

    Args:
        screen_name: 스크린 이름 (PascalCase, 예: Login, Settings)
        feature_dir: feature 디렉토리 경로 (예: login, user/settings)
    """
    snake_name = to_snake_case(screen_name)

    import_line = (
        f"import '../features/{feature_dir}/presentation"
        f"/bloc/{snake_name}_bloc.dart';"
    )

    registration_line = f"  sl.registerFactory(() => {screen_name}Bloc());"

    return f"""{import_line}

// Add to setupDependencies() BLoCs section:
{registration_line}"""


def main():
    if len(sys.argv) < 2:
        print(
            "Usage: flutter_di_template.py <ScreenName> [feature_dir]",
            file=sys.stderr,
        )
        print("Example: flutter_di_template.py Login login", file=sys.stderr)
        print(
            "Example: flutter_di_template.py Settings user/settings",
            file=sys.stderr,
        )
        sys.exit(1)

    screen_name = sys.argv[1]
    snake_name = to_snake_case(screen_name)
    feature_dir = sys.argv[2] if len(sys.argv) > 2 else snake_name

    print(generate_di_registration(screen_name, feature_dir))


if __name__ == "__main__":
    main()

File: scripts/flutter/template/flutter-di-template.mjs
#!/usr/bin/env node
// Flutter DI 등록 템플릿 생성 스크립트
// injection_container.dart에 추가할 BLoC 등록 코드를 생성합니다.

import process from 'node:process';

function toSnakeCase(name) {
  let out = '';
  for (let i = 0; i < name.length; i++) {
    const ch = name[i];
    if (ch >= 'A' && ch <= 'Z' && i > 0) {
      out += '_';
    }
    out += ch.toLowerCase();
  }
  return out;
}

function generateDiRegistration(screenName, featureDir, pkg) {
  const snakeName = toSnakeCase(screenName);

  const importPath = pkg
    ? `package:${pkg}/features/${featureDir}/presentation/bloc/${snakeName}_bloc.dart`
    : `../features/${featureDir}/presentation/bloc/${snakeName}_bloc.dart`;
  const importLine = `import '${importPath}';`;
  const registrationLine = `  sl.registerFactory(${screenName}Bloc.new);`;

  return `${importLine}

// Add to setupDependencies() BLoCs section:
${registrationLine}`;
}

function parseArgs(argv) {
  const opts = { screenName: '', featureDir: '', pkg: '' };
  const rest = [...argv];
  const positional = [];
  while (rest.length > 0) {
    const a = rest.shift();
    if (a === '--package' || a === '-p') {
      opts.pkg = rest.shift() ?? '';
    } else if (a === '-h' || a === '--help') {
      printUsage();
      process.exit(0);
    } else if (a.startsWith('-')) {
      process.stderr.write(`Unknown option: ${a}\n`);
      process.exit(2);
    } else {
      positional.push(a);
    }
  }
  opts.screenName = positional[0] ?? '';
  opts.featureDir = positional[1] ?? '';
  return opts;
}

function printUsage() {
  process.stderr.write('Usage: flutter-di-template.mjs <ScreenName> [feature_dir] [--package <pkg>]\n');
  process.stderr.write('Example: flutter-di-template.mjs Login login --package vocabit_app\n');
  process.stderr.write('Example: flutter-di-template.mjs Settings user/settings --package vocabit_app\n');
}

function main() {
  const opts = parseArgs(process.argv.slice(2));
  if (!opts.screenName) {
    printUsage();
    process.exit(1);
  }
  const featureDir = opts.featureDir || toSnakeCase(opts.screenName);
  console.log(generateDiRegistration(opts.screenName, featureDir, opts.pkg));
}

main();
